src/lib/structures/modelMap.ml

```diff
@@ -30,100 +30,182 @@
+
+  module Map = Map.Make (struct
+      type nonrec t = t
+
+      let compare = compare
     end)
+end
+
+module Graph = struct
+  module M = Map.Make
+      (struct
+        type t = Value.t list [@@deriving ord]
+      end)
+
+  type t = Value.t M.t
+
+  let empty = M.empty
+  let add = M.add
+
+  module Fiber = struct
+    include Set.Make (struct
+        type t = Value.t list [@@deriving ord]
+      end)
+
+    (* For an argument (x_1, ..., x_n) of the function represented by the graph,
+       prints the SMT-LIB formula:
+        (and (= arg_0 x_1)
+          (and (= arg_1 x2)
+            ... (= arg_n x_n).
+    *)
+    let rec pp_args ctr ppf = function
+      | [] -> ()
+      | [arg] ->
+        Fmt.pf ppf "(= arg_%i %a)" ctr Value.pp arg
+      | arg :: args ->
+        Fmt.pf ppf "(and (= arg_%i %a) %a)"
+          ctr
+          Value.pp arg
+          (pp_args (ctr + 1)) args
+
+    (* For a fiber [x; y; z; ...] of the function represented by the graph,
+       prints the SMT-LIB formula:
+        (or (and (= arg_0 x_0) (and (= arg_1 x_1) ...))
+          (or (and (= arg_0 y_0) (and (= arg_1 y_1) ...))
+            ...
+    *)
+    let pp ppf fiber =
+      let rec aux ppf seq =
+        match seq () with
+        | Seq.Nil -> ()
+        | Cons (args, seq) when Stdcompat.Seq.is_empty seq ->
+          Fmt.pf ppf "%a" (pp_args 0) args
+        | Cons (args, seq) ->
+          Fmt.pf ppf "(or %a %a)"
+            (pp_args 0) args
+            aux seq
+      in
+      aux ppf (to_seq fiber)
+  end
+
+  (* Compute all the fibers of the function represented by the graph. *)
+  let inverse graph =
+    M.fold (fun arg_vals ret_val acc ->
+        match Value.Map.find_opt ret_val acc with
+        | Some fib ->
+          Value.Map.add ret_val (Fiber.add arg_vals fib) acc
+        | None ->
+          Value.Map.add ret_val (Fiber.of_list [arg_vals]) acc
+      ) graph Value.Map.empty
+
+  let pp_inverse ppf rel =
+    let rec aux ppf seq =
+      match seq () with
+      | Seq.Nil -> ()
+      | Cons ((ret_val, _), seq) when Stdcompat.Seq.is_empty seq ->
+        Fmt.pf ppf "%a" Value.pp ret_val
+      | Cons ((ret_val, fiber), seq) ->
+        Fmt.pf ppf "(@[<hv>ite %a@ %a@ %a)@]"
+          Fiber.pp fiber
+          Value.pp ret_val
+          aux seq
+    in
+    aux ppf (Value.Map.to_seq rel)
+end
 
-module V = Set.Make
+module P = Map.Make
     (struct
-      type t = (E.t * (X.r * string)) list * (X.r * string)
-      let compare (l1, (v1,_)) (l2, (v2,_)) =
-        let c = X.hash_cmp v1 v2 in
-        if c <> 0 then c
-        else
-          let c = ref 0 in
-          try
-            List.iter2
-              (fun (_,(x,_)) (_,(y,_)) ->
-                 let d = X.hash_cmp x y in
-                 if d <> 0 then begin c := d; raise Exit end
-              ) l1 l2;
-            !c
-          with
-          | Exit -> !c
-          | Invalid_argument _ -> List.length l1 - List.length l2
+      type t = sy [@@deriving ord]
```

Comment:
It feels icky to use `@@deriving ord` to provide the arguments to a functor, it feels very implicit. Defining `let compare = compare_sy` explicitly would be more explicit.